isisd/isis_sr.c

```diff
@@ -0,0 +1,1654 @@
+
+	switch (circuit->circ_type) {
+	case CIRCUIT_T_BROADCAST:
+		ladj_sid = adj->sr.ipv6.ladj_sid;
+		if (ladj_sid) {
+			isis_sr_adj_sid_install_uninstall(
+				false, circuit, adj, ladj_sid->sid, AF_INET6,
+				&adj->ipv6_addresses[0]);
+			isis_tlvs_del_lan_adj_sid(circuit->ext, ladj_sid);
+			adj->sr.ipv6.ladj_sid = NULL;
+		}
+		break;
+	case CIRCUIT_T_P2P:
+		adj_sid = adj->sr.ipv6.adj_sid;
+		if (adj_sid) {
+			isis_sr_adj_sid_install_uninstall(
+				false, circuit, adj, adj_sid->sid, AF_INET6,
+				&adj->ipv6_addresses[0]);
+			isis_tlvs_del_adj_sid(circuit->ext, adj_sid);
+			adj->sr.ipv6.adj_sid = NULL;
+		}
+		break;
+	default:
+		flog_err(EC_LIB_DEVELOPMENT, "%s: unexpected circuit type: %u",
+			 __func__, circuit->circ_type);
+		exit(1);
+	}
+}
+
+#ifndef FABRICD
+static int isis_sr_adj_state_change(struct isis_adjacency *adj)
+{
+	struct isis_circuit *circuit = adj->circuit;
+
+	if (!circuit->area->srdb.enabled)
+		return 0;
+
+	if (adj->adj_state != ISIS_ADJ_UP) {
+		isis_sr_circuit_del_ipv4_adj_sid(circuit, adj);
+		isis_sr_circuit_del_ipv6_adj_sid(circuit, adj);
+	}
+
+	return 0;
+}
+
+static int isis_sr_adj_ipv4_enabled(struct isis_adjacency *adj)
+{
+	struct isis_circuit *circuit = adj->circuit;
+
+	if (!circuit->area->srdb.enabled)
+		return 0;
+
+	isis_sr_circuit_add_ipv4_adj_sid(circuit, adj);
+
+	return 0;
+}
+
+static int isis_sr_adj_ipv4_disabled(struct isis_adjacency *adj)
+{
+	struct isis_circuit *circuit = adj->circuit;
+
+	if (!circuit->area->srdb.enabled)
+		return 0;
+
+	isis_sr_circuit_del_ipv4_adj_sid(circuit, adj);
+
+	return 0;
+}
+
+static int isis_sr_adj_ipv6_enabled(struct isis_adjacency *adj)
+{
+	struct isis_circuit *circuit = adj->circuit;
+
+	if (!circuit->area->srdb.enabled)
+		return 0;
+
+	isis_sr_circuit_add_ipv6_adj_sid(circuit, adj);
+
+	return 0;
+}
+
+static int isis_sr_adj_ipv6_disabled(struct isis_adjacency *adj)
+{
+	struct isis_circuit *circuit = adj->circuit;
+
+	if (!circuit->area->srdb.enabled)
+		return 0;
+
+	isis_sr_circuit_del_ipv6_adj_sid(circuit, adj);
+
+	return 0;
+}
+
+static int isis_sr_if_new_hook(struct interface *ifp)
+{
+	struct isis_circuit *circuit;
+	struct isis_area *area;
+	struct connected *connected;
+	struct listnode *node;
+
+	circuit = circuit_scan_by_ifp(ifp);
+	if (!circuit)
+		return 0;
+
+	area = circuit->area;
+	if (!area)
+		return 0;
+
+	/*
+	 * Update the Node-SID flag of the configured Prefix-SID mappings if
+	 * necessary. This needs to be done here since isisd reads the startup
+	 * configuration before receiving interface information from zebra.
+	 */
+	FOR_ALL_INTERFACES_ADDRESSES (ifp, connected, node) {
+		struct sr_prefix_cfg *pcfg;
+
+		pcfg = isis_sr_cfg_prefix_find(area, connected->address);
+		if (!pcfg)
+			continue;
+
+		if (isis_sr_prefix_is_node_sid(ifp, &pcfg->prefix)
+		    && !CHECK_FLAG(pcfg->flags, ISIS_PREFIX_SID_NODE)) {
+			SET_FLAG(pcfg->flags, ISIS_PREFIX_SID_NODE);
+			lsp_regenerate_schedule(area, area->is_type, 0);
+		}
```

Comment:
Node SID flags is not correctly report in PSID.

Reply:
Good catch. I fixed the issue and now the N-flag is being set appropriately for routes associated to loopback interfaces [3].